hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -855,12 +1080,128 @@ public void breakLease(final Path path, final TracingContext tracingContext) thr
     client.breakLease(getRelativePath(path), tracingContext);
   }
 
-  public void rename(final Path source, final Path destination, TracingContext tracingContext) throws
-          AzureBlobFileSystemException {
+  public void rename(final Path source, final Path destination,
+      final AzureBlobFileSystem azureBlobFileSystem, TracingContext tracingContext) throws
+          IOException {
     final Instant startAggregate = abfsPerfTracker.getLatencyInstant();
     long countAggregate = 0;
     boolean shouldContinue;
 
+    if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
+      LOG.debug("Rename for src: {} dst: {} for non-HNS blob-endpoint",
+          source, destination);
+      final Boolean isSrcExist;
+      final Boolean isSrcDir;
+      /*
+      * Fetch the list of blobs in the given sourcePath.
+      */
+      List<BlobProperty> srcBlobProperties = getListBlobs(source,
+          tracingContext, null, null, true);
+      final BlobProperty blobPropOnSrc;
+      if (srcBlobProperties.size() > 0) {
+        LOG.debug("src {} exists and is a directory", source);
+        isSrcExist = true;
+        isSrcDir = true;
+        /*
+        * Fetch if there is a marker-blob for the source blob.
+        */
+        BlobProperty blobPropOnSrcNullable = getBlobPropertyWithNotFoundHandling(source, tracingContext);
```

Comment:
Why can't we use the same variable blobPropOnSrc here ?

Reply:
We should have a single getBlobProperty call before going to if else block.

Reply:
blobPropOnSrc has to be the final blob related to the source directory marker. There can be chance where there is no marker (in case of dir created by external libraries). In case there is no marker, we will create marker and then save its info in the blobPropOnSrc.